src/execution/window_executor.cpp

```diff
@@ -1428,22 +1428,37 @@ class WindowValueGlobalState : public WindowExecutorGlobalState {
 	WindowValueGlobalState(const WindowExecutor &executor, const idx_t payload_count,
 	                       const ValidityMask &partition_mask, const ValidityMask &order_mask)
 	    : WindowExecutorGlobalState(executor, payload_count, partition_mask, order_mask),
-	      payload_collection(payload_data)
+	      payload_collection(payload_data), ignore_nulls(&no_nulls)
 
 	{
 		if (!arg_types.empty()) {
 			payload_collection.Initialize(Allocator::Get(executor.context), arg_types, payload_count);
 		}
+
+		auto &wexpr = executor.wexpr;
+		if (wexpr.ignore_nulls) {
+			switch (wexpr.type) {
+			case ExpressionType::WINDOW_LEAD:
+			case ExpressionType::WINDOW_LAG:
+			case ExpressionType::WINDOW_FIRST_VALUE:
+			case ExpressionType::WINDOW_LAST_VALUE:
+			case ExpressionType::WINDOW_NTH_VALUE:
+				ignore_nulls = &FlatVector::Validity(payload_collection.chunk.data[0]);
+				break;
+			default:
+				break;
+			}
+		}
 	}
 
 	// The partition values
 	DataChunk payload_data;
 	// The partition values
 	WindowDataChunk payload_collection;
+	// Mask to use for exclusion if we are not ignoring NULLs
+	ValidityMask no_nulls;
 	// IGNORE NULLS
-	ValidityMask ignore_nulls;
-	// Serialisation lock
-	mutex lock;
+	ValidityMask *ignore_nulls;
```

Comment:
nit: `optional_ptr<ValidityMask>`

Reply:
Actually it seems like this is always set - can it be a `reference<ValidityMask>` instead?

Reply:
That would require moving code around. And my apologies for always forgetting about smart pointers!